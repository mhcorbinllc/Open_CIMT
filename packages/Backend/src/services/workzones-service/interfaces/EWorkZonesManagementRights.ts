import {convertDictionaryToViewLabelArray} from "../../../utils/convertDictionaryToViewLabelArray";

export enum EWorkZonesManagementRights {
  WORKZONES_VIEW = "WZM_V",
  WORKZONES_EDIT = "WZM_E",
}

const labelsDic: Record<EWorkZonesManagementRights, string>  = {
  [EWorkZonesManagementRights.WORKZONES_VIEW]: 'CIMs / View',
  [EWorkZonesManagementRights.WORKZONES_EDIT]: 'CIMs / Edit',
};

export const EWorkZonesManagementRightsArray = convertDictionaryToViewLabelArray(labelsDic);
