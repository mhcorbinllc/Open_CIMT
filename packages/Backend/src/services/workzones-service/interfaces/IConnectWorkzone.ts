export interface IConnectWorkzone {
  id: number;
  name: string;
  speedlimit: number;    // In miles
  isWorkzone: true;
  areWorkersPresent: boolean;
  shoulder: string;
  lane: string;
  laneWidth: number;  // Int
  start: string;      // ISO string without ms
  end: string;        // ISO string without ms
  forEver: boolean;
  itisCodes: number[];

  point: {
    type: "Point";
    coordinates: (number | null)[];
  };

  heading: string[];

  broadcastRegion: {
    type: "Polygon";
    coordinates: number[][][];
  };

  usePath: boolean;
  path: {
    type: "LineString";
    coordinates: number[][];
  };

  useRadius: boolean;
  radius: number;

  // Broadcast settings
  serviceChannelStr: string;
  psid: string;
  txmode: string;
  interval: number;
  priority: number;
  useSignature: boolean;
  useEncryption: boolean;
}
