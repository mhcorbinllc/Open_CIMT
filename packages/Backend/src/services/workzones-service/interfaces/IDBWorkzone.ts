import {IWorkzone} from "./IWorkzone";

export interface IDBWorkzone extends IWorkzone {
  dbSearchText: string;
}

export const convertIWorkzoneToIDBWorkzone = (workzone: IWorkzone): IDBWorkzone => {
  const dbWorkzone: IDBWorkzone = {...workzone} as any;
  dbWorkzone.dbSearchText = [
    workzone.workzoneId,
    workzone.kapschId,
    workzone.itisCodes.join(', '),
    workzone.name,
    workzone.notes,
  ]
    .filter(Boolean)
    .join(' ');

  return dbWorkzone;
};

export const convertIDBWorkzoneToIWorkzone = (dbWorkzone: IDBWorkzone): IWorkzone => {
  const workzone: IWorkzone = {...dbWorkzone};
  workzone.start = new Date(workzone.start);
  workzone.end = new Date(workzone.end);
  delete (workzone as any)._id;
  delete (workzone as any).dbSearchText;

  return workzone;
};
