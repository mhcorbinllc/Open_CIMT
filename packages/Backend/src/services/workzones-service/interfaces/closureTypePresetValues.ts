import {
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  IWorkzone,
} from "./IWorkzone";

export const closureTypePresetValues: Record<EWZClosureType, Partial<IWorkzone>> = {
  [EWZClosureType.NONE]: {
    itisCodes: [],
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.LEFT_LANE_CLOSED_MERGE_RIGHT]: {
    itisCodes: "1025 769 12545 8195 771 7451 13579".split(' ').map(Number),
    closedLane: EWZLane.LEFT,
    closedShoulder: EWZShoulder.LEFT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.RIGHT_LANE_CLOSED_MERGE_LEFT]: {
    itisCodes: "1025 769 12545 8196 771 7451 13580".split(' ').map(Number),
    closedLane: EWZLane.RIGHT,
    closedShoulder: EWZShoulder.RIGHT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.LEFT_SHOULDER_CLOSED]: {
    itisCodes: "1025 769 8209 771".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.LEFT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.RIGHT_SHOULDER_CLOSED]: {
    itisCodes: "1025 769 8208 771".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.RIGHT,
    speedLimitInMiles: 0,
  },
  [EWZClosureType.REDUCED_SPEED_15_MPH]: {
    itisCodes: "1025 268 12559 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 15,
  },
  [EWZClosureType.REDUCED_SPEED_20_MPH]: {
    itisCodes: "1025 268 12564 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 20,
  },
  [EWZClosureType.REDUCED_SPEED_25_MPH]: {
    itisCodes: "1025 268 12569 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 25,
  },
  [EWZClosureType.REDUCED_SPEED_30_MPH]: {
    itisCodes: "1025 268 12574 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 30,
  },
  [EWZClosureType.REDUCED_SPEED_35_MPH]: {
    itisCodes: "1025 268 12579 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 35,
  },
  [EWZClosureType.REDUCED_SPEED_40_MPH]: {
    itisCodes: "1025 268 12584 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 40,
  },
  [EWZClosureType.REDUCED_SPEED_45_MPH]: {
    itisCodes: "1025 268 12589 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 45,
  },
  [EWZClosureType.REDUCED_SPEED_50_MPH]: {
    itisCodes: "1025 268 12594 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 50,
  },
  [EWZClosureType.REDUCED_SPEED_55_MPH]: {
    itisCodes: "1025 268 12599 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 55,
  },
  [EWZClosureType.REDUCED_SPEED_60_MPH]: {
    itisCodes: "1025 268 12604 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 60,
  },
  [EWZClosureType.REDUCED_SPEED_65_MPH]: {
    itisCodes: "1025 268 12609 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 65,
  },
  [EWZClosureType.REDUCED_SPEED_70_MPH]: {
    itisCodes: "1025 268 12614 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 70,
  },
  [EWZClosureType.LEFT_TURN_SPEED_15_MPH]: {
    itisCodes: "13594 268 12559 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 15,
  },
  [EWZClosureType.LEFT_TURN_SPEED_20_MPH]: {
    itisCodes: "13594 268 12564 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 20,
  },
  [EWZClosureType.LEFT_TURN_SPEED_25_MPH]: {
    itisCodes: "13594 268 12569 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 25,
  },
  [EWZClosureType.LEFT_TURN_SPEED_30_MPH]: {
    itisCodes: "13594 268 12574 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 30,
  },
  [EWZClosureType.LEFT_TURN_SPEED_35_MPH]: {
    itisCodes: "13594 268 12579 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 35,
  },
  [EWZClosureType.LEFT_TURN_SPEED_40_MPH]: {
    itisCodes: "13594 268 12584 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 40,
  },
  [EWZClosureType.LEFT_TURN_SPEED_45_MPH]: {
    itisCodes: "13594 268 12589 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 45,
  },
  [EWZClosureType.LEFT_TURN_SPEED_50_MPH]: {
    itisCodes: "13594 268 12594 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 50,
  },
  [EWZClosureType.LEFT_TURN_SPEED_55_MPH]: {
    itisCodes: "13594 268 12599 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 55,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_15_MPH]: {
    itisCodes: "13599 268 12559 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 15,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_20_MPH]: {
    itisCodes: "13599 268 12564 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 20,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_25_MPH]: {
    itisCodes: "13599 268 12569 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 25,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_30_MPH]: {
    itisCodes: "13599 268 12574 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 30,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_35_MPH]: {
    itisCodes: "13599 268 12579 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 35,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_40_MPH]: {
    itisCodes: "13599 268 12584 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 40,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_45_MPH]: {
    itisCodes: "13599 268 12589 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 45,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_50_MPH]: {
    itisCodes: "13599 268 12594 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 50,
  },
  [EWZClosureType.RIGHT_TURN_SPEED_55_MPH]: {
    itisCodes: "13599 268 12599 8720".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 55,
  },
  [EWZClosureType.HEIGHT_LIMIT_14_FT_0_IN]: {
    itisCodes: "2574 13569 12712 8709".split(' ').map(Number),
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,
  },
};
