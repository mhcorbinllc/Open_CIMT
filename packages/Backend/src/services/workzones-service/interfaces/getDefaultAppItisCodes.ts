import {IItisCode} from "./IWorkzonesAppSettings";

export const getDefaultAppItisCodes = (): IItisCode[] => {
  return [
    {
      code: 1025,
      label: "Road Construction",
    },
    {
      code: 769,
      label: "Closed to Traffic",
    },
    {
      code: 12545,
      label: "1",
    },
    {
      code: 8196,
      label: "Right Lane",
    },
    {
      code: 8195,
      label: "Left Lane",
    },
    {
      code: 771,
      label: "Closed Ahead",
    },
    {
      code: 7451,
      label: "Merge",
    },
    {
      code: 13580,
      label: "Left",
    },
    {
      code: 13579,
      label: "Right",
    },
    {
      code: 8208,
      label: "Right Shoulder",
    },
    {
      code: 8209,
      label: "Left Shoulder",
    },
    {
      code: 13594,
      label: "Left Turn",
    },
    {
      code: 13599,
      label: "Right Turn",
    },
    {
      code: 268,
      label: "Speed Limit",
    },
    {
      code: 12559,
      label: "15",
    },
    {
      code: 12564,
      label: "20",
    },
    {
      code: 12569,
      label: "25",
    },
    {
      code: 12574,
      label: "30",
    },
    {
      code: 12579,
      label: "35",
    },
    {
      code: 12584,
      label: "40",
    },
    {
      code: 12589,
      label: "45",
    },
    {
      code: 12594,
      label: "50",
    },
    {
      code: 12599,
      label: "55",
    },
    {
      code: 12604,
      label: "60",
    },
    {
      code: 12609,
      label: "65",
    },
    {
      code: 12614,
      label: "70",
    },
    {
      code: 8720,
      label: "MPH",
    },
    {
      code: 2574,
      label: "Height Limit",
    },
    {
      code: 13569,
      label: "Ahead",
    },
    {
      code: 12712,
      label: "168",
    },
    {
      code: 8709,
      label: "Inches",
    },
  ];
};
