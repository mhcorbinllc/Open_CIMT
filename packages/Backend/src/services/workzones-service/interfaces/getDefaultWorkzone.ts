import * as moment from "moment";

import {
  EWZBroadcastSelectionMode,
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  IWorkzone,
} from "./IWorkzone";
import {
  EWZClosureSelectionMode,
} from "./IWorkzonesAppSettings";

export const getDefaultWorkzone = (): IWorkzone => {
  const now = moment().seconds(0);

  return {
    workzoneId: '',
    kapschId: -1,
    name: '',
    active: true,

    notes: '',
    start: now.toDate(),
    end: now.clone().add(3, 'days')
      .toDate(),
    forEver: false,

    closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
    closureType: EWZClosureType.NONE,
    itisCodes: [],
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,

    closedLaneWidthInMeters: 0,
    workersPresent: false,

    point: {
      lat: 0,
      lng: 0,
      alt: 0,
    },

    path: {points: []},

    heading: [],

    broadcast: {
      selectionMode: EWZBroadcastSelectionMode.REGION,
      selectionByRegion: {region: {points: []}},
      selectionByRadius: {radiusInMeters: 0},
    },

    createdAt: 0,
    updatedAt: 0,
    deletedAt: 0,
  };
};

