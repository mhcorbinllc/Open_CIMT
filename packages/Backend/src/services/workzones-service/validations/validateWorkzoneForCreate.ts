import {IWorkzone} from "../interfaces/IWorkzone";

import {IValidateDataEngineRules} from "core-library/dist/commonJs/validation-engine";
import {validateDataMethods} from "core-library/dist/commonJs/validation-engine";
import {validateWorkzoneForUpdate} from "./validateWorkzoneForUpdate";

// This is very limited validation, we check only security and if the props exist and have correct type.
// We don't validate the logic or if the values are right since we allow the user , this is done on broadcast.

export const validateWorkzoneForCreate: IValidateDataEngineRules<IWorkzone> = {
  ...validateWorkzoneForUpdate,
  workzoneId: (workzoneId: string) => {
    if (!validateDataMethods.isString(workzoneId)) return 'Should be a string';
    if (workzoneId.trim()) return 'Should not provide workzoneId on Create';
    return true;
  },
};
