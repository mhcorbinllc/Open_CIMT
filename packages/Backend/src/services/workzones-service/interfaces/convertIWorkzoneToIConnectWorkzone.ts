import {IWorkzone} from "./IWorkzone";
import {IConnectWorkzone} from "./IConnectWorkzone";
import {IWorkzonesAppSettings} from "./IWorkzonesAppSettings";

export const convertIWorkzoneToIConnectWorkzone = (
  {
    workzone,
    workzoneAppSettings,
  }: {
    workzone: IWorkzone;
    workzoneAppSettings: IWorkzonesAppSettings;
  },
): IConnectWorkzone => {
  return {
    id: workzone.kapschId,
    name: workzone.name,
    speedlimit: Math.round(workzone.speedLimitInMiles),
    isWorkzone: true,
    areWorkersPresent: workzone.workersPresent,
    shoulder: workzone.closedShoulder,
    lane: workzone.closedLane,
    laneWidth: Math.round(workzone.closedLaneWidthInMeters),
    start: dateToIsoString(workzone.start),
    end: dateToIsoString(workzone.end),
    forEver: workzone.forEver,
    itisCodes: workzone.itisCodes,

    // Point
    point: {
      type: "Point",
      coordinates: [
        workzone.point.lng,
        workzone.point.lat,
        workzone.point.alt === undefined
          ? null
          : workzone.point.alt,
      ],
    },

    heading: workzone.heading,

    // Path
    usePath: !!workzone.path.points.length,
    path: {
      type: "LineString",
      coordinates:
        workzone.path.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
    },

    // Selection mode: Region
    broadcastRegion: {
      type: "Polygon",
      coordinates: [
        workzone.broadcast.selectionByRegion.region.points
          .map(point => ([
            point.lng,
            point.lat,
            point.alt || 0,
          ])),
      ],
    },

    // Selection mode: Radius
    useRadius: !!workzone.broadcast.selectionByRadius.radiusInMeters,
    radius: workzone.broadcast.selectionByRadius.radiusInMeters,

    // CIMs features options
    serviceChannelStr: workzoneAppSettings.serviceChannel,
    psid: workzoneAppSettings.psid,
    txmode: workzoneAppSettings.txmode,
    interval: workzoneAppSettings.intervalInMs,
    priority: workzoneAppSettings.priority,
    useSignature: workzoneAppSettings.useSignature,
    useEncryption: workzoneAppSettings.useEncryption,
  };
};

const dateToIsoString = (date: Date | string | number): string => {
  return new Date(date).toISOString()
    .split('.')[0] + "Z";
};
