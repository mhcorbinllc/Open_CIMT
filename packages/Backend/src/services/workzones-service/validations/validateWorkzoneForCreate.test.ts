import {
  EWZHeading,
  IWorkzone,
} from "../interfaces/IWorkzone";

import {getDefaultWorkzone} from "../interfaces/getDefaultWorkzone";
import {validationEngine} from "core-library/dist/commonJs/validation-engine";

import {validateWorkzoneForCreate} from "./validateWorkzoneForCreate";

describe('validateWorkzoneForCreate', () => {
  test('Missing workzoneId (correct)', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...getDefaultWorkzone(),
      name: 'Demo CIM',
    }));

    workzone.heading = [EWZHeading.EAST];

    const validation = validationEngine(workzone, validateWorkzoneForCreate);

    expect(validation.isValid).toBeTruthy();
    expect(validation).toMatchSnapshot();
  });

  test('Providing workzoneId (wrong)', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...getDefaultWorkzone(),
      name: 'Demo CIM',
      workzoneId: 'ed-98467204',
    }));

    workzone.heading = [EWZHeading.EAST];

    const validation = validationEngine(workzone, validateWorkzoneForCreate);

    expect(validation.isValid).toBeFalsy();
    expect(validation.messages.length).toBe(1);
    expect(validation).toMatchSnapshot();
  });
});
