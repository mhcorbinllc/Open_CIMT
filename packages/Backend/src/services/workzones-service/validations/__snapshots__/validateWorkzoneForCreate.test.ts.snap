// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateWorkzoneForCreate Missing workzoneId (correct) 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForCreate Providing workzoneId (wrong) 1`] = `
Object {
  "dataValidation": Object {
    "workzoneId": "Should not provide workzoneId on Create",
  },
  "isValid": false,
  "messages": Array [
    "Workzone id: Should not provide workzoneId on Create",
  ],
}
`;
